src/webapp/components/form/FormState.ts

```diff
@@ -81,50 +85,215 @@ export type FormFieldState =
+    sectionsState: FormSectionState[]
+): FormSectionState[] {
+    return sectionsState.map(section => {
+        if (section?.subsections) {
+            return {
+                ...applyEffectNotApplicableFieldUpdatedInSection(section),
+                subsections: applyEffectNotApplicableFieldUpdatedInSections(section?.subsections),
+            };
+        } else {
+            return applyEffectNotApplicableFieldUpdatedInSection(section);
+        }
+    });
+}
+
+function applyEffectNotApplicableFieldUpdatedInSection(
+    sectionState: FormSectionState
+): FormSectionState {
+    return {
+        ...sectionState,
+        fields: sectionState.fields.map(field => {
+            if (field.hasNotApplicable) {
+                const notApplicableField = sectionState.fields.find(
+                    f => f.notApplicableFieldId === field.id
+                );
+
+                // TODO: FIXME TypeScript error returning the corresponding fieldValue type
+                const fieldValue = (
+                    notApplicableField?.value ? getEmptyValueForField(field) : getFieldValue(field)
+                ) as any;
+
+                return {
+                    ...field,
+                    value: fieldValue,
+                    disabled: !!notApplicableField?.value,
+                };
+            }
+            return field;
+        }),
+    };
+}
+
 export function updateFormState(prevFormState: FormState, updatedField: FormFieldState): FormState {
     return {
         ...prevFormState,
         sections: updateSectionsState(prevFormState.sections, updatedField),
     };
 }
 
-export function updateSectionsState(
+export function updateFormStateWithFieldErrors(
+    formState: FormState,
+    updatedField: FormFieldState,
+    fieldValidationErrors: ValidationError[]
+): FormState {
+    return {
+        ...formState,
+        sections: updateSectionsState(formState.sections, updatedField, fieldValidationErrors),
+    };
+}
+
+function updateSectionsState(
     prevFormSectionsState: FormSectionState[],
-    updatedField: FormFieldState
+    updatedField: FormFieldState,
+    fieldValidationErrors?: ValidationError[]
 ): FormSectionState[] {
-    const updatedSections = prevFormSectionsState.map(section => {
+    return prevFormSectionsState.map(section => {
         if (section?.subsections) {
             return {
-                ...updateSectionState(section, updatedField),
-                subsections: updateSectionsState(section?.subsections, updatedField),
+                ...updateSectionState(section, updatedField, fieldValidationErrors),
+                subsections: updateSectionsState(
+                    section?.subsections,
+                    updatedField,
+                    fieldValidationErrors
+                ),
             };
         } else {
-            return updateSectionState(section, updatedField);
+            return updateSectionState(section, updatedField, fieldValidationErrors);
```

Comment:
we need to update a section only if the updatedField belong to that section right? Can we add that condition? Currently we are calling update section code for all the sections unnecessarily.